### ISSUE TITLE
expired keys in stick-table removed only after "show table <name>"

### ISSUE THREAD

**Issue by zstyblik** [haproxy/haproxy#428]
Hello,

I'm observing rather strange behaviour after upgrade to HAproxy v2.0.10. `show table` command reports that stick-table is almost full, resp. utilization of stick-table is over 90%. However, when `show table <name>` is executed, then expired keys are evicted and `used` count is updated to real usage.

```
root@example.com:~# while true; do date; echo "show table" | socat stdio /run/haproxy/haproxy1.sock; sleep 60; done
Fri Dec 27 12:11:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1927773
# table: my_bk1_sticky, type: string, size:5242880, used:1932360

Fri Dec 27 12:12:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1927841
# table: my_bk1_sticky, type: string, size:5242880, used:1932423

Fri Dec 27 12:13:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1927895
# table: my_bk1_sticky, type: string, size:5242880, used:1932367

Fri Dec 27 12:14:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1927958
# table: my_bk1_sticky, type: string, size:5242880, used:1932331

Fri Dec 27 12:15:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1928028
# table: my_bk1_sticky, type: string, size:5242880, used:1932348

Fri Dec 27 12:16:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1928093
# table: my_bk1_sticky, type: string, size:5242880, used:1932337

Fri Dec 27 12:17:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1928163
# table: my_bk1_sticky, type: string, size:5242880, used:1932337

Fri Dec 27 12:18:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1928223
# table: my_bk1_sticky, type: string, size:5242880, used:1932291

Fri Dec 27 12:19:37 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:1928294
# table: my_bk1_sticky, type: string, size:5242880, used:1932305

^C
root@example.com:~# echo "show table my_bk2_sticky" | socat stdio /run/haproxy/haproxy1.sock > stick_table.out
root@example.com:~# echo "show table" | socat stdio /run/haproxy/haproxy1.sock
Fri Dec 27 12:20:13 UTC 2019
# table: my_bk2_sticky, type: string, size:2097152, used:247158
# table: my_bk1_sticky, type: string, size:5242880, used:1932389
```

When I inspected `stick_table.out` it more or less correlated with usage reported before and after. Meaning, there were a lot of `exp=0` keys which were pruned when `show table <name>` has been executed.

This is from different server, but configuration and issue is the same:

```
root@example.com:~# cat stick_table.out | grep -v -e 'exp=0' | wc -l
228493
root@example.com:~# cat stick_table.out | grep -e 'exp=0' | wc -l
1762526
```

## Output of `haproxy -vv` and `uname -a`

```
root@example.com:~# haproxy -vv
HA-Proxy version 2.0.10-1~bpo9+1 2019/11/26 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.0.10=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.0k  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.22 2016-07-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace

root@example.com:~# uname -a
Linux someHost 4.19.0-0.bpo.2-amd64 #1 SMP Debian 4.19.16-1~bpo9+1 (2019-02-07) x86_64 GNU/Linux
```

## What's the configuration?

```
global
    log    /dev/log len 2048 local0
    stats socket /var/run/haproxy/haproxy1.sock process 1 mode 770 level admin expose-fd listeners
    server-state-file /var/run/haproxy/servers_state
[...]
peers my_peers
    peer example01 1.2.3.1:1024
    peer example02 1.2.3.2:1024
    peer example03 1.2.3.3:1024
    peer example04 1.2.3.4:1024
    peer example05 1.2.3.5:1024
[...]
backend my_bk1_sticky
    stick on path
    stick-table type string len 128 size 5m expire 1d peers my_peers

backend my_bk2_sticky
   # This one is affected for sure
   stick on req.hdr("some-header") if   { hdr(some-header) -m found }
    stick on req.hdr("another-header")         if ! { hdr(another-header) -m found }
    stick-table type string len 64 size 2m expire 1d peers my_peers
```

## Steps to reproduce the behavior

This is only a guess on how to reproduce this.

1. create `frontend`, create `backend` with `stick-table`. To speed things up, `stick-table` should be reasonably small and expiration reasonably short.
2. send traffic through
3. observe stats via control socket `echo "show table" | socat stdio /run/haproxy/haproxy1.sock`
4. `echo "show table my_bk2_sticky" | socat stdio /run/haproxy/haproxy1.sock` should lower count of `used`, resp. prune `exp=0` keys from the table

## Actual behavior

Count of `used` entries in stick-table keeps growing despite keys have already expired. Count of `used` entries is updated and keys evicted from the table when `echo "show table my_bk2_sticky" | socat stdio /run/haproxy/haproxy1.sock` is executed.

## Expected behavior

Count of `used` entries reported by `show table` is accurate/up-to-date. It doesn't have to be real-time accurate, but it shouldn't be updated only when contents of stick-table are shown.

## Do you have any idea what may have caused this?

I haven't observed this before v2.0.10, so maybe some bug or regression? I see some changes/fixes related to peers in v2.0.10 CHANGELOG. Could one of those be behind this strange behaviour? On the other hand, issue manifested itself ~ one week after upgrade.
Also, HAproxy is being reloaded even on upgrade. No idea what happens with stick-table data. 

## Do you have an idea how to solve the issue?

Unfortunately, no. Not yet anyway.

**Comment by wtarreau** [comment]
Hello,

thanks for your detailed report. I'm really puzzled as I have no idea what could possibly cause this, and I cannot reproduce it, at best I can catch very few transient "exp=0" that are currently expiring but even then it's rare:

```
$ socat - /tmp/sock1 <<< "show table http" > table.out
$ grep -c exp=0 table.out 
21
$ grep -c -v exp=0 table.out 
189678
```

I also tried to roll back to exactly 2.0.10 but no effect. I don't see what in the configuration could cause such a behavior to happen so I think you've hit a bug provoked by a corner case. Does it happen immediately when you start haproxy or only after some time? (your timeouts are long so I guess it's at least after one day but could you trigger it with lower intervals?). Does it also happen if you remove your peers section (in case you can reproduce it on a non-prod platform) ? Are you sure it happens on the machine taking the traffic and not maybe on another one ? Would it be possible that entries that remain stuck there were learned from peers only (probably you can check a few of them in your logs to see if they come from the local machine, or if you find them on all machines but one).

What I *suspect* is that some entries are inserted but the stick-table timeout not re-armed. For example we could imagine that entries learned from the peers wouldn't always re-arm the stick-table's timer and that if the node doesn't take any traffic, this timer would never be re-armed. That's just a wild guess of course, but we can try to see if the responses to the questions above tend to support this theory or not.

**Comment by zstyblik** [comment]
Hi @wtarreau,

thank you for your reply and apologies for late reply on my part. I wasn't feeling well past couple days.

> thanks for your detailed report.

Actually, there are couple things to add just in case:

1. While debugging this issue, I've noticed "couple of" old(some really old) haproxy processes are still hanging around. This can be caused by various out-of-scope reasons, eg signal got lost, bogus connection, bug etc.
2. Issues related to systemd and D-Bus. Could these affect haproxy in any way?
3. This issue happens only in production, so far, and only on one set of servers with specific use. There is another set which has same setup as far as stick-tables go, but different use.
And just to make it clear, these two sets don't share `peers` and can't talk with each other.
4. On the other hand, there were holidays, therefore staging environment has seen next to nil use. Also, staging is running at v2.0.12 now, was v2.0.11 when I reported this issue.

> I'm really puzzled as I have no idea what could possibly cause this, and I cannot reproduce it, at best I can catch very few transient "exp=0" that are currently expiring but even then it's rare:

I'm not surprised you cannot reproduce it. I didn't expect it to be easy(I think).

> I also tried to roll back to exactly 2.0.10 but no effect. I don't see what in the configuration could cause such a behavior to happen so I think you've hit a bug provoked by a corner case.

> Does it happen immediately when you start haproxy or only after some time? 

I estimate this started approximately one week after update to v2.0.10.

> Does it also happen if you remove your peers section (in case you can reproduce it on a non-prod platform) ?

I haven't tried it and I cannot do that in production. Unfortunately, this doesn't happen in staging as stated above.

> Are you sure it happens on the machine taking the traffic and not maybe on another one ?

It is happening on all servers in that set and traffic goes through all of them.

> Would it be possible that entries that remain stuck there were learned from peers only

But shouldn't they be removed when `exp=0`? Is there some thread which should service this? Would it be possible there is so much "churn" it's not able to keep up?

Also, if my understanding is correct, this issue shouldn't be issue at all since `exp=0` keys in table are replaced if needed. And the only issue here is that output of `show table` is incorrect.

> (probably you can check a few of them in your logs to see if they come from the local machine, or if you find them on all machines but one).

I can dump stick-table at all servers and compare it. If you have any tips on how to compare these, please, let me know as I was struggling with it at some point since I've learned this is sort of async.

Anyway, I've found dump at one of the servers when I reported this issue and did something rather naive:

```
% cat dump.out | grep -e 'exp=0' | grep -e 'server_id=1' | wc -l
355932
% cat dump.out | grep -e 'exp=0' | grep -v -e 'server_id=1' | wc -l
1406594
```
But I'm not sure whether `server_id` is origin of the stick-table entry or not.

> What I _suspect_ is that some entries are inserted but the stick-table timeout not re-armed. For example we could imagine that entries learned from the peers wouldn't always re-arm the stick-table's timer and that if the node doesn't take any traffic, this timer would never be re-armed. That's just a wild guess of course, but we can try to see if the responses to the questions above tend to support this theory or not.

I'm sorry, but this is beyond my understanding :|

Here is a bit expanded config file:

```
global
    log    /dev/log len 2048 local0
    stats socket /var/run/haproxy/haproxy1.sock process 1 mode 770 level admin expose-fd listeners
    server-state-file /var/run/haproxy/servers_state
[...]
peers my_peers
    peer example01 1.2.3.1:1024
    peer example02 1.2.3.2:1024
    peer example03 1.2.3.3:1024
    peer example04 1.2.3.4:1024
    peer example05 1.2.3.5:1024
[...]

frontend www
   [... probably your typical FE config ...]
   default_backend my_bk1_sticky
   [... and so on ...]
   use_backend bk_foo if something

frontend proxy_bk2
    bind unix@/var/lib/haproxy/proxy_bk2.sock user haproxy group varnish mode 0660
    default_backend my_bk2_sticky

backend my_bk1_sticky
    stick on path
    stick-table type string len 128 size 5m expire 1d peers my_peers

backend my_bk2_sticky
   # This one is affected for sure
   stick on req.hdr("some-header") if   { hdr(some-header) -m found }
    stick on req.hdr("another-header")         if ! { hdr(another-header) -m found }
    stick-table type string len 64 size 2m expire 1d peers my_peers
```

Actually, `my_bk2_sticky` is used only through Unix socket and nowhere else. On the other hand, `my_bk1_sticky` is used only in typical http 80/443 frontend section.

Last, but not least. I feel a bit stupid, because I did not try to reproduce it. Yes, I've provided idea how to reproduce it, but I haven't tried to reproduce it in eg. VM or docker.
Also, I can upgrade to v2.0.12 now and we can see whether it makes any difference.
And is there anything or any "admin" commands or strace which could shed some light into what's going on and why?

Thank you.

**Comment by wtarreau** [comment]
>     1. While debugging this issue, I've noticed "couple of" old(some really old) haproxy processes are still hanging around. This can be caused by various out-of-scope reasons, eg signal got lost, bogus connection, bug etc.
> 
>     2. Issues related to systemd and D-Bus. Could these affect haproxy in any way?

It could be anything including long sessions with missing timeouts. If you're running with a master socket enable, you can connect to it and bounce to these older processes to check what they're doing
and if they're still processing sessions, for example:
```
  # show proc
  @-2 show sess
```
It would be interesting to see if they're in fact dealing with zombie peers connections!

> I'm not surprised you cannot reproduce it. I didn't expect it to be easy(I think).

OK, thanks for confirming :-)

> I estimate this started approximately one week after update to v2.0.10.

And do you remember what version you upgraded from, that didn't exhibit this issue ?
 
> > Does it also happen if you remove your peers section (in case you can reproduce it on a non-prod platform) ?
> 
> I haven't tried it and I cannot do that in production. Unfortunately, this doesn't happen in staging as stated above.

OK, no problem.

> > Would it be possible that entries that remain stuck there were learned from peers only
> 
> But shouldn't they be removed when `exp=0`?

They should obviously but since they don't, I'm trying to figure in what case they manage to escape this hard rule. And given that I can't reproduce the issue, I'm guessing it's in some corner case situations.

> Is there some thread which should service this?

Actually it's a task but that's the same principle. It is scheduled to wake up at the earliest next expiration for each stick table.

> Would it be possible there is so much "churn" it's not able to keep up?

No because it purges all expired entries in one run. Actually the only ones that are not removed are the ones still being tracked ("use" > 0). But it's not what happens here because you said they disappear after a "show table", which apparently forces the purge. Actually there is a reason why "show table" can clear an entry: it needs to hold a reference to the entry while it's dumping it, and when it drops its reference it must check if the refcount becomes zero indicating that during this time the entry expired. So in your case as a side effect of dropping the reference it evicts expired entries. This really indicates that the purge task doesn't run because all conditions are met to purge these entries.

> Also, if my understanding is correct, this issue shouldn't be issue at all since `exp=0` keys in table are replaced if needed. And the only issue here is that output of `show table` is incorrect.

No, it also eats more memory because expired entries will only be reclaimed if the table is full.

> > (probably you can check a few of them in your logs to see if they come from the local machine, or if you find them on all machines but one).
> 
> I can dump stick-table at all servers and compare it. If you have any tips on how to compare these, please, let me know as I was struggling with it at some point since I've learned this is sort of async.

I would proceed like this: on each machine:
```
  # socat /path/to/socket - <<< "show table my_bk2_sticky" | cut -f2- table.out -d' ' > table-$HOSTNAME.out
```

Then you can simply run:
```
  # diff -y -w 160 table-host1.out table-host2.out
  ...
```
Then it's interesting to look for "exp=0" in the output there and see what other nodes have.

> Anyway, I've found dump at one of the servers when I reported this issue and did something rather naive:
> 
> ```
> % cat dump.out | grep -e 'exp=0' | grep -e 'server_id=1' | wc -l
> 355932
> % cat dump.out | grep -e 'exp=0' | grep -v -e 'server_id=1' | wc -l
> 1406594
> ```
> 
> But I'm not sure whether `server_id` is origin of the stick-table entry or not.

No idea either but at least it indicates that this can happen with regular stickiness.

> Last, but not least. I feel a bit stupid, because I did not try to reproduce it. Yes, I've provided idea how to reproduce it, but I haven't tried to reproduce it in eg. VM or docker.

Well, you don't have to feel stupid for this :-)  And even if you manage to reproduce it in such an environment, it doesn't imply that anyone else will. It looks complicated enough and can depend on a lot of parameters.

> Also, I can upgrade to v2.0.12 now and we can see whether it makes any difference.

This will at least rule out some possible bugs though I don't imagine which ones could cause the problem.

> And is there anything or any "admin" commands or strace which could shed some light into what's going on and why?

You can use "show peers" to see if any peer is stuck or to see the number of entries learned from other ones, but this will not provide much information beyond some hints I'm afraid. I think the diff will help more actually.

**Comment by zstyblik** [comment]
> It could be anything including long sessions with missing timeouts. If you're running with a master socket enable, you can connect to it and bounce to these older processes to check what they're doing
> and if they're still processing sessions, for example:
> 
> ```
>   # show proc
>   @-2 show sess
> ```

This is a nice trick.

> It would be interesting to see if they're in fact dealing with zombie peers connections!

Where I could, I did take a look. Indeed, connections/backends without timeouts which will never ever timeout. This was sort of expected at places I could inspect it. Unfortunately, I couldn't inspect all places, because master socket is not present despite being configured. Perhaps it was introduced later(and I don't necessarily mean haproxy) and reloads don't spawn it. I don't know as I lack knowledge in multiple directions here.

Also, fixing systemd didn't change anything related to this issue, although haproxy needs to be restarted. And I'm fairly sure this would make the issue to go away.

> And do you remember what version you upgraded from, that didn't exhibit this issue ?

v2.0.9, at least that's what history says.

> > But shouldn't they be removed when `exp=0`?
> 
> They should obviously but since they don't, I'm trying to figure in what case they manage to escape this hard rule. And given that I can't reproduce the issue, I'm guessing it's in some corner case situations.
> 
> > Is there some thread which should service this?
> 
> Actually it's a task but that's the same principle. It is scheduled to wake up at the earliest next expiration for each stick table.

Just a note, it seems that reload schedules it as I've noticed drop in stick-table utilization after reload, but it doesn't make the issue to go away. There is a side effect, though, as it seems to refresh expiration of keys on that server, so I've ended up with:

```
# Reloaded server
key=X use=0 exp=76166894 server_id=1 server_name=be01
# Peer which has not been reloaded
key=X use=0 exp=0 server_id=1 server_name=be01
```

Hm, very weak idea here, but could network issues/timeout somehow cause re-learning and re-refresh(rearm) of stick-table entries. I mean can be seen above, but without reload of haproxy. 

> > Would it be possible there is so much "churn" it's not able to keep up?
> 
> No because it purges all expired entries in one run. Actually the only ones that are not removed are the ones still being tracked ("use" > 0). But it's not what happens here because you said they disappear after a "show table", which apparently forces the purge. Actually there is a reason why "show table" can clear an entry: it needs to hold a reference to the entry while it's dumping it, and when it drops its reference it must check if the refcount becomes zero indicating that during this time the entry expired. So in your case as a side effect of dropping the reference it evicts expired entries. This really indicates that the purge task doesn't run because all conditions are met to purge these entries.

I see.

> > Also, if my understanding is correct, this issue shouldn't be issue at all since `exp=0` keys in table are replaced if needed. And the only issue here is that output of `show table` is incorrect.
> 
> No, it also eats more memory because expired entries will only be reclaimed if the table is full.

Actually, that's what I meant :)

> > I can dump stick-table at all servers and compare it. If you have any tips on how to compare these, please, let me know as I was struggling with it at some point since I've learned this is sort of async.
> 
> I would proceed like this: on each machine:
> 
> ```
>   # socat /path/to/socket - <<< "show table my_bk2_sticky" | cut -f2- table.out -d' ' > table-$HOSTNAME.out
> ```
> 
> Then you can simply run:
> 
> ```
>   # diff -y -w 160 table-host1.out table-host2.out
>   ...
> ```
> 
> Then it's interesting to look for "exp=0" in the output there and see what other nodes have.
>

I did, but there are too many to check. Each file is 100-150MB. However, this seems to be interesting:

```
# server05
key=X use=0 exp=69518317 server_id=4 server_name=be04
# server02
key=X use=0 exp=0 server_id=1 server_name=be01
```
I see some discrepancies here. First of all, `exp`. And second, `server_name`.

But then, there are entries which are the same, expired, but still present:
```
key=Y use=0 exp=0 server_id=4 server_name=be04
key=Y use=0 exp=0 server_id=4 server_name=be04
```

And some are missing?:

```
key=Z use=0 exp=0 server_id=4 server_name=be04
<
```

Also, one more thing. Would it be possible there is a race condition when two or more peers add the same key into the table, but with different `sever_name`, resp. route? In other words, split-brain scenario.

> > Also, I can upgrade to v2.0.12 now and we can see whether it makes any difference.
> 
> This will at least rule out some possible bugs though I don't imagine which ones could cause the problem.

It's up to you, resp. how much you want (me) to debug it further.

> You can use "show peers" to see if any peer is stuck or to see the number of entries learned from other ones, but this will not provide much information beyond some hints I'm afraid. I think the diff will help more actually.

Here's a brief output as I don't exactly know what to look for. I have checked management guide, but all I can understand is `status`. Feel free to point me in some direction(including source code) and I'll get there ;)
I can post full output of `show peers` from one or all servers, if needed. 

Speaking of status, how come 2(**EDIT**: corrected, was 1) out of 5 servers has connection established to itself, but others don't?

```
server01.txt:  0x556ba1729210: id=server05(remote) addr=X:1024 status=ESTA reconnect=1s confirm=0 tx_hbt=0 rx_hbt=552 no_hbt=0 new_conn=3387704 proto_err=0
server01.txt:  0x556ba1728370: id=server04(remote) addr=X:1024 status=ESTA reconnect=0s confirm=0 tx_hbt=0 rx_hbt=2180 no_hbt=0 new_conn=3387961 proto_err=0
server01.txt:  0x556ba17274d0: id=server03(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=0 rx_hbt=18 no_hbt=0 new_conn=3386068 proto_err=0
server01.txt:  0x556ba1726650: id=server02(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=3386339 proto_err=0
server01.txt:  0x556ba1723a60: id=server01(local)  addr=X:1024 status=NONE reconnect=<NEVER> confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=0 proto_err=0
---
server02.txt:  0x563453313210: id=server05(remote) addr=X:1024 status=ESTA reconnect=1s confirm=0 tx_hbt=0 rx_hbt=40 no_hbt=0 new_conn=1241122709 proto_err=446368
server02.txt:  0x563453312370: id=server04(remote) addr=X:1024 status=ESTA reconnect=1s confirm=0 tx_hbt=39470 rx_hbt=43 no_hbt=4 new_conn=1242228138 proto_err=1117159
server02.txt:  0x5634533114f0: id=server03(remote) addr=X:1024 status=ESTA reconnect=4s confirm=0 tx_hbt=9863 rx_hbt=8717 no_hbt=0 new_conn=382491165 proto_err=129858
server02.txt:  0x563453310200: id=server02(local)  addr=X:1024 status=ESTA reconnect=<NEVER> confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=0 proto_err=45903068
server02.txt:  0x56345330da50: id=server01(remote) addr=X:1024 status=ESTA reconnect=1s confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=1235483127 proto_err=1135319
---
server03.txt:  0x555d9be17210: id=server05(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=0 rx_hbt=736 no_hbt=0 new_conn=963537480 proto_err=71
server03.txt:  0x555d9be16390: id=server04(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=0 rx_hbt=304 no_hbt=0 new_conn=964154861 proto_err=77
server03.txt:  0x555d9be150a0: id=server03(local)  addr=X:1024 status=ESTA reconnect=<NEVER> confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=0 proto_err=0
server03.txt:  0x555d9be14200: id=server02(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=8717 rx_hbt=9860 no_hbt=0 new_conn=108855619 proto_err=83
server03.txt:  0x555d9be11a50: id=server01(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=11240 rx_hbt=917237 no_hbt=0 new_conn=29645735 proto_err=34622
---
server04.txt:  0x55dcd2bbc270: id=server05(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=786 rx_hbt=5173 no_hbt=0 new_conn=885378680 proto_err=311389
server04.txt:  0x55dcd2bbaf80: id=server04(local)  addr=X:1024 status=NONE reconnect=<NEVER> confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=0 proto_err=0
server04.txt:  0x55dcd2bba0e0: id=server03(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=309 rx_hbt=0 no_hbt=2 new_conn=963223084 proto_err=1138011
server04.txt:  0x55dcd2bb9240: id=server02(remote) addr=X:1024 status=ESTA reconnect=3s confirm=0 tx_hbt=0 rx_hbt=39468 no_hbt=0 new_conn=959888089 proto_err=16814
server04.txt:  0x55dcd2bb6a90: id=server01(remote) addr=X:1024 status=ESTA reconnect=4s confirm=0 tx_hbt=97130 rx_hbt=1245 no_hbt=2 new_conn=876835069 proto_err=1004561
---
server05.txt:  0x55fcaf2d8de0: id=server05(local)  addr=X:1024 status=NONE reconnect=<NEVER> confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=0 proto_err=0
server05.txt:  0x55fcaf2d7f40: id=server04(remote) addr=X:1024 status=ESTA reconnect=4s confirm=0 tx_hbt=5177 rx_hbt=786 no_hbt=2 new_conn=887750344 proto_err=743543
server05.txt:  0x55fcaf2d70a0: id=server03(remote) addr=X:1024 status=ESTA reconnect=2s confirm=0 tx_hbt=736 rx_hbt=0 no_hbt=0 new_conn=961862579 proto_err=1137857
server05.txt:  0x55fcaf2d6200: id=server02(remote) addr=X:1024 status=ESTA reconnect=1s confirm=0 tx_hbt=0 rx_hbt=0 no_hbt=0 new_conn=962420238 proto_err=692728
server05.txt:  0x55fcaf2d3a50: id=server01(remote) addr=X:1024 status=ESTA reconnect=4s confirm=0 tx_hbt=93749 rx_hbt=1244 no_hbt=28 new_conn=875710202 proto_err=1004375
```

Thank you for bearing with me :)

**Comment by zstyblik** [comment]
Hello,

I just want to let you know that it seems like issue is gone since upgrade to v2.0.13. Of course, it's possible something else (external) has changed, but I think 2.0.13 has somehow fixed this.

**Comment by wtarreau** [comment]
OK,thanks a lot for this useful feedback. I'm now closing the issue, but do not hesitate to reopen it if the problem resurfaces.
